### PR TITLE
Added: support delay start for room cleaning

### PR DESCRIPTION
**Pull request by ileodo**
## Current Behaviour
- Once user toggle one Room switch, robot will start the room based cleaning just for that room

## Behaviour after this PR
- While user toggling the Room switches, cleaner will not take any action. 
- After selecting the rooms want to be cleaned by toggling the Room switches, user can start the room based clean by turn on the fan.

Thanks to the original PR
https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/pull/200

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -123,6 +123,8 @@ class XiaomiRoborockVacuum {
     this.connectRetry = setTimeout(() => void 0, 100); // Noop timeout only to initialise the property
     this.getStateInterval = setInterval(() => void 0, GET_STATE_INTERVAL_MS); // Noop timeout only to initialise the property
 
+    this.roomIdsToClean = [];
```
Comment (ileodo) [inline index.js:126]: I think it make sense to have this array as a state of the robot.

File: index.js [hunk 2]
```diff
@@ -905,26 +930,31 @@ class XiaomiRoborockVacuum {
...
+  async setCleaningRoom(state, roomId) {
     await this.ensureDevice("setCleaning");
 
     try {
-      if (state && !this.isCleaning) {
-        // Start cleaning
+      if (state && !this.isCleaning && !this.isPaused) {
         this.log.info(
-          `ACT setCleaning | ${this.model} | Start cleaning Room ID ${room}, not charging.`
+          `ACT setCleaningRoom | ${this.model} | Enable cleaning Room ID ${roomId}.`
```
Comment (ileodo) [inline index.js:939]: correct the log category

File: index.js [hunk 3]
```diff
@@ -1034,17 +1064,19 @@ class XiaomiRoborockVacuum {
     this.log.info(
       `INF createRoom | ${this.model} | Room ${roomName} (${roomId})`
     );
-    this.services[roomName] = new Service.Switch(
+
+    this.services.rooms = this.services.rooms || {};
```
Comment (ileodo) [inline index.js:1068]: group all the room switch services under `this.services.rooms`

**Comment by ileodo** [comment]
@afharo could you please have a quick look?

File: index.js [hunk 4]
```diff
@@ -905,26 +930,31 @@ class XiaomiRoborockVacuum {
...
-        await this.device.cleanRooms([room]);
-      } else if (!state) {
-        // Stop cleaning
+        if (!this.roomIdsToClean.includes(roomId)){
+          this.roomIdsToClean.push(roomId)
+        }
+      } else if (!state && !this.isCleaning && !this.isPaused){
         this.log.info(
-          `ACT setCleaning | ${this.model} | Stop cleaning and go to charge.`
+          `ACT setCleaningRoom | ${this.model} | Diable cleaning Room ID ${roomId}.`
```
Comment (afharo) [inline index.js:946]:
```suggestion
          `ACT setCleaningRoom | ${this.model} | Disable cleaning Room ID ${roomId}.`
```

File: miio/lib/devices/viomivacuum.js [hunk 5]
```diff
@@ -158,6 +158,10 @@ module.exports = class extends Vacuum.with(
     ).then(checkResult);
   }
 
+  resumeCleanRooms(listOfRooms) {
+    return this.pause();
```
Comment (afharo) [inline miio/lib/devices/viomivacuum.js:162]: Is it `this.pause()`? Is calling pause actually unpausing the cleaning? 🤔

**Review by afharo — COMMENTED** [review]
Thank you @ileodo, it is looking great!

I added a couple of minor comments that I think it will improve performance and future maintainability

**Comment by afharo** [comment]
@ileodo can I ask you another change, please? Do you mind adding a section in the README explaining how it works?

File: miio/lib/devices/viomivacuum.js [hunk 6]
```diff
@@ -159,7 +159,8 @@ module.exports = class extends Vacuum.with(
   }
 
   resumeCleanRooms(listOfRooms) {
-    return this.pause();
+    // TODO need to find out the miio command here.
```
Comment (ileodo) [inline miio/lib/devices/viomivacuum.js:162]: Not sure what's the miio command for viomivacuum. @afharo could you please help with this?

File: index.js [hunk 7]
```diff
@@ -1420,12 +1461,18 @@ class XiaomiRoborockVacuum {
   getServices() {
     if (this.config.delay) this.sleep(5000);
     this.log.debug(`DEB getServices | ${this.model}`);
-    return Object.keys(this.services).map((key) => {
+    return Object.keys(this.services).reduce((services, key) => {
       if (key !== "fan" && this.services.fan.addLinkedService) {
         this.services.fan.addLinkedService(this.services[key]);
```
[the remark this answers is not in the document]
Comment (ileodo) [reply index.js]: nice catch

File: miio/lib/devices/viomivacuum.js (hunk 6 above)
Comment (afharo) [reply miio/lib/devices/viomivacuum.js:162]: I'm not sure either. Let's run the same command as the start room cleaning for now.

File: index.js [hunk 8]
```diff
@@ -1349,10 +1379,17 @@ class XiaomiRoborockVacuum {
     await this.ensureDevice("setPauseState");
 
     try {
-      if (state) {
-        await this.device.activateCleaning();
-      } else {
+      if (state && this.isPaused) {
+        if(this.roomIdsToClean.size > 0) {
+          await this.device.resumeCleanRooms();
```
Comment (afharo) [inline index.js:1384]: Should we send the list of rooms as an argument to this function?

**Review by afharo — APPROVED** [review]
It looks great! Thank you @ileodo for the effort on this PR!

I'll release it straight away as 0.15.0

**Comment by normen** [comment]
Meh.. before this I could go "Hey Siri, turn on vacuum (room name)", now I have to go "Hey Siri, turn on vacuum (room name)" "Hey Siri, vacuum to 100%".

If there was an auto-timeout for the fan it would be much nicer, I could go "Hey Siri, turn on vacuum bedroom and vacuum hallway".

**Comment by afharo** [comment]
Hi @normen,

I like that suggestion. We can build on top of this essential piece of work @ileodo did to support multiple-room cleaning. Can you create a PR that adds this behaviour?

Maybe we can add a config value that allows users to overwrite the timer (some users may need 10s to make sure all the room buttons they are after are enabled).

Another approach could be to set an automation: _When room switch A, B or C are enabled, trigger a clean_. This way, you can say "Hey Siri, turn on vacuum A and vacuum B", Siri will enable both switches and the automation will automatically trigger the clean button.

**Comment by normen** [comment]
> I like that suggestion. We can build on top of this essential piece of work @ileodo did to support multiple-room cleaning. Can you create a PR that adds this behaviour?

Done, see https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/pull/326

I am still getting another error now with the latest version 0.15.1 though, see #324

**Comment by thechris1992** [comment]
> Meh.. before this I could go "Hey Siri, turn on vacuum (room name)", now I have to go "Hey Siri, turn on vacuum (room name)" "Hey Siri, vacuum to 100%".
> 
> If there was an auto-timeout for the fan it would be much nicer, I could go "Hey Siri, turn on vacuum bedroom and vacuum hallway".

then a 10? sec Timeout (or specific Timeout) after enable a room cleaning option would be the fix.

**Comment by normen** [comment]
> then a 10? sec Timeout (or specific Timeout) after enable a room cleaning option would be the fix.

Yeah, thats why I added that feature :)

**Comment by thechris1992** [comment]
yeha but the room cleaning doesn't start after enable the room to clean

**Comment by normen** [comment]
> yeha the room cleaning doesn't start after enable the room to clean

Yeah, you need to set the `roomTimeout` variable..?

**Comment by thechris1992** [comment]
> > yeha the room cleaning doesn't start after enable the room to clean
> 
> Yeah, you need to set the `roomTimeout` variable..?

done this with 1000 seems like this is in seconds. with 5 its very better and works. 
This should be documented at the main page ^^ (Optional parameters & Instruction)

Thanks!

**Comment by normen** [comment]
> This should be documented at the main page ^^ (Optional parameters & Instruction)

It is 🙃

**Comment by thechris1992** [comment]
oh right on the button. 
my bad

**Pull request merged**